### PR TITLE
[SPARK-15125][SQL] Changing CSV data source mapping of empty quoted strings in the data to empty strings instead of null

### PR DESCRIPTION
**Pull request by sureshthalamati**
## What changes were proposed in this pull request?

Currently  empty quoted strings in the input CSV file are incorrectly recognized as null value.  This patch adds fix to recognizes quoted empty strings(eg; 1, "")  in the data are recognized by default as empty string values.  

New CSV  option emptyValue is added to allows users  to specify the value to write to the output CSV file for empty string, and also the value in in an input file  that should be interpreted as empty string in addition to the empty quoted string.

DATA :
col1,col2
1,"-"
2,""
3,
4,"A

**Default Old behaviour:**
scala> val df = spark.read.format("csv").option("nullValue", "-").option("header", true).load("/Users/suresht/sparktests/emptystring.csv")
df: org.apache.spark.sql.DataFrame = [col1: string, col2: string]

scala> df.show
+----+----+
|col1|col2|
+----+----+
|   1|null|
**|   2|null|**
|   3|null|
|   4|   A|
+----+----+

**Default New behavior :**

scala> val df = spark.read.format("csv").option("nullValue", "-").option("header", true).load("/Users/suresht/sparktests/emptystring.csv")
df: org.apache.spark.sql.DataFrame = [col1: string, col2: string]

scala> df.show
+----+----+
|col1|col2|
+----+----+
|   1|null|
**|   2|    |**
|   3|null|
|   4|   A|
+----+----+

**Example using the emptyValue option:**

val df = spark.createDataFrame(Seq((1, "")))
df.write.format("csv").option("emptyValue", "EMPTY").save("/tmp/data1")

**cat part-r-00000-8d867267-c291-4277-9951-a8b969c0a4d8.csv** 
1,EMPTY

scala> spark.read.format("csv").option("emptyValue", "EMPTY").load("/tmp/data1").show
+---+---+
|_c0|_c1|
+---+---+
|  1|   |
+---+---+
## How was this patch tested?

Added new unit tests to the CSVSuite. 

@falaki @rxin

### CODE REVIEW

**Comment by rxin** [comment]
Shouldn't we always infer these as empty strings, and then users can do a simple project to turn them into nulls?

**Comment by rxin** [comment]
Basically I'm asking why is this config needed if we just treat them as empty strings?

**Comment by sureshthalamati** [comment]
Thanks for reviewing the PR, Reynold.  Only reason I added the option is to keep the current behavior by default.  I agree with you we can treat them as empty strings always.  I am also not aware of any scenario where user wants to treat empty strings as null values. 

If there is no need to keep the current default behavior , I will update the PR.  Please let me know

**Comment by rxin** [comment]
Yea let's update it.

cc @HyukjinKwon @falaki

**Comment by HyukjinKwon** [comment]
+1 for treating them as empty strings

I guess this will be conflicted with https://github.com/apache/spark/pull/12921 because that PR deals with some bug fixes in inferring/converting `null`. I hope this can be merged first because that PR includes 50ish more tests from `HadoopFsRelationTest`.

**Comment by sureshthalamati** [comment]
Thank you for the feedback , Reynold , HyukjinKwon.  Update the PR.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala [hunk 1]
```diff
@@ -555,4 +558,37 @@ class CSVSuite extends QueryTest with SharedSQLContext with SQLTestUtils {
...
     assert(numbers.count() == 8)
   }
+
+  test("load data with empty quoted string fields.") {
+    val results = sqlContext
+      .read
+      .format("csv")
+      .options(Map(
+        "header" -> "true",
+        "nullValue" -> null,
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala]: Could I ask what happen if we don't set `nullValue`?

**Comment by HyukjinKwon** [comment]
@rxin @sureshthalamati Do you mind holding off this change until #12921 is merged? That PR also handles `nullValue`. Apparently, I guess `nullValue` could affect this behaviour.

**Comment by HyukjinKwon** [comment]
Here is what I think CSV datasource should handle `""`, empty string and `nullValue`.

With the option, `nullValue` set to `"null"`, I think 

```csv
,"","null"
```

should produce the records as below:

```
Row(null, "", null)
```

Would this make sense?

**Comment by HyukjinKwon** [comment]
In case of writing, I think

```
Row("", "null", null)
```
should produce the CSV as below:

1. With the option, `nullValue` set to `"null"`, I think 

  ```csv
,null,null
```

2. Without any options, I think 

  ```csv
,null,
```

3. With the option, `nullValue` set to `""`, I think 

  ```csv
,null,
```

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala (hunk 1 above)
Comment (sureshthalamati) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala]: If nullValue is not set  it  will return empty string for null values by default.  The reason I set it explicitly is to make sure my fix is working.  Before my fix it was  retruning null for the empty quoted string , and empty  string for null values by default.

**Comment by sureshthalamati** [comment]
I am not sure what was the history behind returning empty String for null value. In my opinion it should be null be default.  current behavior is also  inconsistent; for numerics it will return null and for strings it will return empty string by default. 

Example:
See the Year (int), and comment (String in the following data).
year,make,model,comment,price
2017,Tesla,Mode 3,looks nice.,35000.99
,Chevy,Bolt,,29000.00
2015,Porsche,"",,
scala> val df= sqlContext.read.format("csv").option("header", "true").option("inferSchema", "true").load("/tmp/test1.csv")
df: org.apache.spark.sql.DataFrame = [year: int, make: string ... 3 more fields]

scala> df.show
+----+-------+------+-----------+--------+
|year|   make| model|    comment|   price|
+----+-------+------+-----------+--------+
|2017|  Tesla|Mode 3|looks nice.|35000.99|
|null|  Chevy|  Bolt|           | 29000.0|
|2015|Porsche|  null|           |    null|
+----+-------+------+-----------+--------+


I can update this PR to change the nullValue default if  needed,

**Comment by HyukjinKwon** [comment]
+1 for @sureshthalamati  
#12921 handles the inconsistent behaviour and this is why I think we should hold off this until that PR is merged.

**Comment by HyukjinKwon** [comment]
@andrewor14 Here it seems it is concluded that `""` is a string and a empty string is `null`. Because `""` is a legitimate string, this can be a field name whereas a empty string might not be able (`null`). Although we cannot query to empty strings, we can still see when it performs `show()`.

It seems anyway it is decided that empty strings will be unnamed columns with prefix (`_c`). Do you think this can be generalized to data sources and Spark SQL including fields name with spaces (eg. `"    "`)?

**Comment by HyukjinKwon** [comment]
After the discussion here, https://github.com/apache/spark/pull/12904, then I think 

```scala
StructField("", StringType, nullable = true)
```

should be renamed to `_c0`.

**Comment by HyukjinKwon** [comment]
After the discussion https://github.com/apache/spark/pull/13041, I think the field below:

```scala
StructField("", StringType, nullable =true)
```

should be renamed to a field having prefix `_c`.

**Comment by sureshthalamati** [comment]
@HyukjinKwon  does your previous comment for meant for  some other PR ? This PR does not have any change you mentioned above. Am I missing some thing ?

@andrewor14

**Comment by HyukjinKwon** [comment]
@sureshthalamati oh, the comments are not related with this PR but moving the discussion to here was suggested. So, i did. Sorry for that if it was confusing.

**Comment by sureshthalamati** [comment]
Thanks for the input @falaki . Sorry for the delated reply, sone how I missed the notifications. I will update the patch with  option and also add test case  for 17916.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVOptions.scala [hunk 2]
```diff
@@ -90,6 +90,7 @@ private[csv] class CSVOptions(@transient private val parameters: Map[String, Str
   val permissive = ParseModes.isPermissiveMode(parseMode)
 
   val nullValue = parameters.getOrElse("nullValue", "")
+  val emptyValue = parameters.getOrElse("emptyValue", "")
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVOptions.scala]: When `nullValue` and `emptyValue` are both `""` in default, don't they conflict?

Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVOptions.scala]: +1 for setting a explicit precedence for both.

**Comment by HyukjinKwon** [comment]
FWIW, it seems `read.csv()` in R seems not differenciating it from `""`.

```r
> bt <- "A,B,C,D
+ ,\"\",20"
>
> b<- read.csv(text=bt, na.strings=c(""))
> b
   A  B  C  D
1 NA NA 20 NA
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVOptions.scala (hunk 2 above)
Comment (sureshthalamati) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVOptions.scala]: Yes, null and empty can not be differentiated when they are set to same value.  Currently null value check has higher precedence than the empty value. 

input.csv
1,
2,”” 

Output will be: 
1, null
2, null


I think this behavior  is ok.  By default  Univocity CSV parser  used in spark  also returns  null for empty strings.

I agree we should document this behavior.

**Comment by felixcheung** [comment]
hmm, @HyukjinKwon the behavior with your R example is because of the `na.strings=c("")` parameter, which tells it to treat empty `""` as `NA` (== JVM null in Spark)

**Comment by HyukjinKwon** [comment]
Oh, sure it is true. I set `na.strings` as `""` because the equivalent option, `nullValue` has the same default value in Spark's CSV :).

**Comment by antoniobarbuzzi** [comment]
I agree with @HyukjinKwon and comment https://github.com/apache/spark/pull/12904#issuecomment-217605160, except for third use case, i.e. I'd rather do this:

With the option, nullValue set to `""`
`"","a",` should be converted to `Row("", "a", null)`

In fact, just like the univocity parser does [here](https://github.com/uniVocity/univocity-parsers/blob/master/src/main/java/com/univocity/parsers/csv/CsvParserSettings.java#L63-L72):
> if the parser does not read any character from the input, and the input is within quotes , the empty is used instead of an empty string


Regarding the PR, I'd deletegate the handling of this to the univocity parser, setting the `settings.setEmptyValue(params.emptyValue)` in the univocity's CSVSettings [here](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVParser.scala#L34) and [here](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVParser.scala#L77) too.

**Comment by sureshthalamati** [comment]
I was testing the fix with different scenarios mentioned in the comments. I can not make 
CSV writer write quoted  empty string for empty strings in the data.  One of the issue I filed  got fixed , but still can not make it work.   

https://github.com/uniVocity/univocity-parsers/issues/123

**Comment by gatorsmile** [comment]
What is the status of this PR?

**Comment by gatorsmile** [comment]
@falaki Should we continue this PR?

**Comment by MaxGekk** [comment]
The issue has been already solved by https://github.com/apache/spark/commit/7a2d4895c75d4c232c377876b61c05a083eab3c8 . The PR can be closed.

**Pull request closed without merging**